src/mesh/mesh_adaptation.cpp

```diff
@@ -0,0 +1,119 @@
+#include "mesh_adaptation.h"
+
+namespace PHiLiP {
+
+template <int dim, typename real, typename MeshType>
+MeshAdaptation<dim,real,MeshType>::MeshAdaptation(double critical_res_input, int total_ref_cycle, double refine_frac, double coarsen_frac)
+    : critical_residual(critical_res_input)
+    , total_refinement_cycles(total_ref_cycle)
+    , refinement_fraction(refine_frac)
+    , coarsening_fraction(coarsen_frac)
+    , current_refinement_cycle(0)
+    {}
+
+template <int dim, typename real, typename MeshType>
+int MeshAdaptation<dim,real,MeshType>::adapt_mesh(std::shared_ptr< DGBase<dim, real, MeshType> > dg)
+{
+    if (!refine_mesh)
+    {
+        return 0;
+    }
+
+    if (total_refinement_cycles == current_refinement_cycle)
+    {
+        return 0;
+    }
+
+    cellwise_errors.reinit(0);
+    cellwise_errors.reinit(dg->high_order_grid->triangulation->n_active_cells());
+    compute_cellwise_errors(dg);
```

Comment:
Keep in mind the Single Responsibility Principle. Currently, your class is doing 2 big things:
1. Computing the error estimates
2. Refining
Consider separating the error estimate into a separate class whose single responsibility is to compute errors.
Then pass this error estimate to this mesh adaptation class whose single responsibility is to refine our DG class.

Reply:
Error estimation is now done by a separate class and has a general virtual function for future extension.